### DISCUSSION TITLE
Saving crops

### DISCUSSION THREAD

**Discussion by TimSousa1** [NVlabs/instant-ngp#335] [Q&A]
Hey! Just wanted to know if there's a way to save what I set in the “Crop aabb” parameter so that when I run testbed.exe and load that scene it only trains on that specific region 
Or do I have to set it every time?

**Comment by Tom94** [comment]
Hi there, you can use the Python bindings, which are able to control almost any parameter that the GUI exposes.

For your use case, you can modify the portion of `run.py` that runs before the training loop to set
```python
testbed.render_aabb = ngp.BoundingBox([min_x, min_y, min_z], [max_x, max_y, max_z])
```

**Reply by leastbad** [reply]
I'm trying to provide literal values as you suggested (just to get/verify it working) but the GUI is not picking up the values when the testbed loads. Can you suggest what line in `run.py` you'd place the code before? Need a rubber duck check...

```python
testbed.render_aabb = ngp.BoundingBox([0.085, -0.255, 0.409], [0.765, 1.228, 0.722])
```

**Reply by dailafing** [reply]
I too am trying to achieve this, so I can save the crop values.
The gui often crashes, and I'm loosing a lot of time re-cropping.
Please could someone let me know where this code should be stored in the run.py file. thanks

@leastbad Did you manage to get this working? Thanks

**Reply by initialvisuals** [reply]
same issue here, running it on a 1660 super if I try to mesh a bit too high or save a snapshot while training it can crash. I don't really understand why saving a snapshot doesn't save all of your parameters? Should be relatively easy for them to have a feature for that?

**Reply by francescofugazzi** [reply]
You can add thise pieces of code inside testbed.cu but keep in mind that the msgpack you will save will not work on the regular code anymore and it's only for crop dimensions and doesn't account the rotation.
Almost at the end of the of the file find the save_snapshot section:
```cuda
void Testbed::save_snapshot(const std::string& filepath_string, bool include_optimizer_state) {
```

and just after 
```cuda
        snapshot["bounding_radius"] = m_bounding_radius;`
```
add

```cuda
	snapshot["crop_aabb"] = m_render_aabb;
	snapshot["edit_aabb"] = m_edit_render_aabb;
```

This will save the crop parameters.

Now you have add the lines of code for being able to load the new parameters in the load_snapshot section:
```cuda
void Testbed::load_snapshot(const std::string& filepath_string) {
```
and after 
```cuda
                m_nerf.training.counters_rgb.measured_batch_size_before_compaction = snapshot["nerf"]["rgb"]["measured_batch_size_before_compaction"];`
```
add

```cuda
		if (m_data_path.empty() && snapshot["nerf"].contains("dataset")) {
			m_nerf.training.dataset = snapshot["nerf"]["dataset"];
			load_nerf();
		} else {
			if (snapshot["nerf"].contains("aabb_scale")) {
				m_nerf.training.dataset.aabb_scale = snapshot["nerf"]["aabb_scale"];
			}
		}
```
And then recompile it

Note: last time I used these lines of code was on early august. I didn't check if they still works on newer revisions. Let me know
